### ISSUE TITLE
Async functions in pbjs.cmd queue

### ISSUE THREAD

**Issue by Zer0Divis0r** [prebid/Prebid.js#6700]
## Type of issue
Feature request

## Description
`$$PREBID_GLOBAL$$.cmd.push` to detect and handle async functions, and run individual queue steps synchronously relative to each other.
I wonder if it would make sense to run the steps in queue consequentely, one after another, using `await`. 

For example: 

```
pbjs.cmd.push(async function() {  /* await for some asynchronous stuff # 1, or run a Promise */  } );
pbjs.cmd.push(async function() {  /* await for some asynchronous stuff # 2, or run a Promise */  } );
```

I'd expect the `stuff # 2` procedure to happen only after `stuff # 1` fulfills.
Right now it would run the two functions in parallel, and not guarantee any execution order.

WDYT?

**Comment by patmmccann** [comment]
Possible duplicate of #3058 , a 6.0 change

**Comment by patmmccann** [comment]
Closing as duplicate

**Comment by patmmccann** [comment]
fwiw we use pbjs.que, not pbjs.cmd https://github.com/prebid/Prebid.js/blob/54d225674f26446fddee1a85103e179fd98e4bc4/src/events.js#L62

**Comment by muuki88** [comment]
Can you describe the use case for this?

From my experience assuming order of execution is rarely a good idea in browserland.

What we do is this

```ts
const pbjsReady: Promise<void> = new Promise(resolve => pbjs.que.push(resolve))
```

and use this promise where we need to do things. IMHO handling this on the user side is a better idea.

**Comment by Zer0Divis0r** [comment]
@muuki88 Your example creates a Promise that fulfulls when the whole queue finishes executing. But that does not provide an option to create a queue step that will hold the whole queue untill that step is fulfilled.

Consider the example from the initial post: there is an asynchronous work that has to be done **in order to proceed** with the rest of the queue. 

Specific examples of use cases: 
- wait for some Ajax data needed for setting bidders config
- wait for ad slots to fill or return empty ads
- stall the queue for debugging to inspect race conditions

**Comment by muuki88** [comment]
Thanks for the clarification. :blush: 

TL;DR: IMHO the `que` should be as simple as possible. It executes the callbacks. It's not a place for synchronization.

There are two options interacting with the `pbjs.que` object on a site (AFAIK)

1. Single controller - One vendor provides, configures and runs prebid.
2. Shared instance - Two or more vendors use a shared prebid instance to run auctions on the site

## Single controller

You have full control over what goes into `pbjs.que` and what not. This means, IMHO, that you don't need the que to synchronize work, but your code should do this. Your example

> wait for some Ajax data needed for setting bidders config

is a perfect example. Assuming the que would handle promises and stall the next execution once the promises resolves or rejects:

```ts
// fetch config settings
pbjs.que.push(() => fetch("..."));

// fetch other stuff you require
pbjs.que.push(() => fetch("..."));

// run auction
pbjs.que.push(() => runAuction());
```

This already requires you to run the `push` calls in right order, otherwise you may run an auction without the other stuff. And this creates numerous pitfalls, wenn you don't properly handle the promises and configure timeouts, e.g.

```
Promise
  // make sure long requests don't block the entire auction
  .race([fetch("..."), createTimeoutPromise(1000))
  // make sure that a single failed promise doesn't crash the entire auction
  .catch(() => ())
```

You can do the same right now with

```
const pbjsReady: Promise<void> = new Promise(resolve => pbjs.que.push(resolve));
const data1: Promise<SomeData> =  fetch("...").then(r => r.json());
const data2: Promise<SomeData> =  fetch("...").then(r => r.json());

await pbjsReady;
await data1;
await data2

// do what every you want to do with the data
```

I personally prefer `Promise.all`, `Promise.race` and `.then` concats, which makes it clearer what's running in parallel and whats sequential. You may noticed that this code already includes an optimization.

```ts
const pbjsReady: Promise<void> = new Promise(resolve => pbjs.que.push(resolve));

const data1 = await  fetch("...").then(r => r.json());
const data2 = await  fetch("...").then(r => r.json());

await pbjsReady;
```

This would run `data1` and `data2` sequentially, which is probably not want you want.


## Shared Controller

This would mean you are at the mercy of the other vendors that they don't block the que with their `promises` that  can take _forever_ . Currently the only way to block the que, is to run expensive operations in the callback function, which I think is rather uncommon. But IO (e.g. requests to an http endpoint) can be common and those can take from milliseconds to minutes.

## Use cases

Regarding your examples

> wait for some Ajax data needed for setting bidders config

I discussed this in the first example. If you need to sync stuff, I would do this in your code. A command que is no place to sync work. Order of execution is guaranteed. If you spawning promises than fullfil sometime in the future this is not the business of the command que IMHO.

> wait for ad slots to fill or return empty ads

How does the que help you here? `pbjs.event`, `gpt.addEventListener` or whatever ad server you use, provides you with the means to react on those things. :smiley: 

> stall the queue for debugging to inspect race conditions

The browser dev tools do provide this. JS is after all single threaded. And if you stop the world with a debugger, you can't debug a race condition. Probably I'm missing something here :sweat_smile: 


I would strongly  advice against this feature as it adds a bunch of new error scenarios that prebid must handle then.

**Comment by patmmccann** [comment]
excellent write up @muuki88
